python/dolfinx/mesh.py

```diff
@@ -80,89 +102,64 @@ def locate_entities_boundary(mesh: cpp.mesh.Mesh,
+    "interval": CellType.interval,
+    "triangle": CellType.triangle,
+    "quadrilateral": CellType.quadrilateral,
+    "tetrahedron": CellType.tetrahedron,
+    "hexahedron": CellType.hexahedron
 }
 
 _meshtags_types = {
-    numpy.int8: cpp.mesh.MeshTags_int8,
-    numpy.int32: cpp.mesh.MeshTags_int32,
-    numpy.int64: cpp.mesh.MeshTags_int64,
-    numpy.double: cpp.mesh.MeshTags_double
+    np.int8: _cpp.mesh.MeshTags_int8,
+    np.int32: _cpp.mesh.MeshTags_int32,
+    np.int64: _cpp.mesh.MeshTags_int64,
+    np.double: _cpp.mesh.MeshTags_double
 }
 
 
 def refine(mesh, cell_markers=None, redistribute=True):
     """Refine a mesh"""
     if cell_markers is None:
-        mesh_refined = cpp.refinement.refine(mesh, redistribute)
+        mesh_refined = _cpp.refinement.refine(mesh, redistribute)
     else:
-        mesh_refined = cpp.refinement.refine(mesh, cell_markers, redistribute)
+        mesh_refined = _cpp.refinement.refine(mesh, cell_markers, redistribute)
 
     coordinate_element = mesh._ufl_domain.ufl_coordinate_element()
     domain = ufl.Mesh(coordinate_element)
-    domain._ufl_cargo = mesh_refined
-    mesh_refined._ufl_domain = domain
-    return mesh_refined
+    return Mesh.from_cpp(mesh_refined, domain)
 
 
 def create_mesh(comm, cells, x, domain,
-                ghost_mode=cpp.mesh.GhostMode.shared_facet,
-                partitioner=cpp.mesh.partition_cells_graph):
-    """Create a mesh from topology and geometry data"""
+                ghost_mode=GhostMode.shared_facet,
+                partitioner=_cpp.mesh.partition_cells_graph):
+    """Create a mesh from topology and geometry arrays"""
     ufl_element = domain.ufl_coordinate_element()
     cell_shape = ufl_element.cell().cellname()
     cell_degree = ufl_element.degree()
-    cmap = cpp.fem.CoordinateElement(_uflcell_to_dolfinxcell[cell_shape], cell_degree)
+    cmap = _cpp.fem.CoordinateElement(_uflcell_to_dolfinxcell[cell_shape], cell_degree)
     try:
-        mesh = cpp.mesh.create_mesh(comm, cells, cmap, x, ghost_mode, partitioner)
+        mesh = _cpp.mesh.create_mesh(comm, cells, cmap, x, ghost_mode, partitioner)
     except TypeError:
-        mesh = cpp.mesh.create_mesh(comm, cpp.graph.AdjacencyList_int64(numpy.cast['int64'](cells)),
-                                    cmap, x, ghost_mode, partitioner)
-
-    # Attach UFL data (used when passing a mesh into UFL functions)
+        mesh = _cpp.mesh.create_mesh(comm, _cpp.graph.AdjacencyList_int64(np.cast['int64'](cells)),
+                                     cmap, x, ghost_mode, partitioner)
     domain._ufl_cargo = mesh
-    mesh._ufl_domain = domain
-    return mesh
+    return Mesh.from_cpp(mesh, domain)
 
 
-def MeshTags(mesh, dim, indices, values):
+def MeshTags(mesh: Mesh, dim: int, indices, values):
 
     if isinstance(values, int):
-        values = numpy.full(indices.shape, values, dtype=numpy.int32)
+        values = np.full(indices.shape, values, dtype=np.int32)
```

Comment:
This cast is not safe according to numpy:

`np.can_cast(int, np.int32)`

Should avoid doing hidden unsafe casts.

Reply:
Cast from int to np.int64 is safe, possible alternative?

Reply:
This PR didn't change the details of this code. Could you create a separate PR if it needs to be changed? There is already a lot going on in this PR.

Reply:
Will do.